proof_of_stake/src/lib.rs

```diff
@@ -2230,69 +2455,54 @@ where
             );
             Some(ValidatorSetUpdate::Consensus(ConsensusValidator {
                 consensus_key,
-                bonded_stake: cur_stake.into(),
+                bonded_stake: new_stake.into(),
             }))
         });
-    let cur_below_capacity_validators =
-        below_capacity_validator_set_handle().at(&next_epoch);
-    let prev_below_capacity_vals =
-        below_capacity_validator_set_handle().at(&current_epoch);
-
-    let below_capacity_validators = cur_below_capacity_validators
-        .iter(storage)
-        .unwrap()
+
+    let prev_consensus_validators = prev_consensus_validator_handle
+        .iter(storage)?
         .filter_map(|validator| {
             let (
                 NestedSubKey::Data {
-                    key: cur_stake,
+                    key: _prev_stake,
                     nested_sub_key: _,
                 },
                 address,
             ) = validator.unwrap();
-            let cur_stake = token::Amount::from(cur_stake);
 
-            tracing::debug!(
-                "Below-capacity validator address {address}, stake {cur_stake}"
-            );
+            let new_state = validator_state_handle(&address)
+                .get(storage, next_epoch, params)
+                .unwrap();
 
-            let prev_validator_stake = validator_deltas_handle(&address)
-                .get_sum(storage, current_epoch, params)
-                .unwrap()
-                .map(token::Amount::from_change)
-                .unwrap_or_default();
+            let prev_validator_stake =
+                read_validator_stake(storage, params, &address, current_epoch)
+                    .unwrap()
+                    .unwrap_or_default();
             let prev_tm_voting_power = into_tm_voting_power(
                 params.tm_votes_per_token,
                 prev_validator_stake,
             );
 
-            // If the validator previously had no voting power, it wasn't in
-            // tendermint set and we have to skip it.
-            if prev_tm_voting_power == 0 {
-                tracing::debug!(
-                    "skipping validator update {address}, it's inactive and \
-                     previously had no voting power"
+            // If the validator is still in the Consensus set, we accounted for
+            // it in the `new_consensus_validators` iterator above
+            if matches!(new_state, Some(ValidatorState::Consensus)) {
+                return None;
+            } else if params.validator_stake_threshold
+                == token::Amount::default()
+                && prev_tm_voting_power == 0
```

Comment:
We can make `prev_tm_voting_power` lazy to avoid the computation if the prev conditions short-circuit

Reply:
added in [19f5aa3](https://github.com/anoma/namada/pull/1576/commits/19f5aa34e26f0f82a7c57846143cf3486543e44b)